### PR TITLE
MOTECH-2573: Added basic auth to IHE

### PR DESCRIPTION
**Pull request by mkruszynski**

### CODE REVIEW

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/HL7Recipient.java [hunk 1]
```diff
@@ -9,12 +9,25 @@
...
 
+    private String recipientUsername;
+
+    private String recipientPassword;
+
     public HL7Recipient() {
     }
 
     public HL7Recipient(String recipientName, String recipientUrl) {
         this.recipientName = recipientName;
```
Comment (sebbrudzinski) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/HL7Recipient.java]: Make it call this(args) with proper parameters. It's a good practice to chain constructors, so they do not repeat the same assignemnts over and over again

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/service/IHETemplateService.java [hunk 2]
```diff
@@ -14,4 +16,12 @@
...
      */
     void sendTemplateToRecipientUrl(String url, String template) throws IOException;
+
+    /**
+     * Sends given CDA template to HL7 recipient with Basic Authentication.
+     *
+     * @param recipient Recipient which will receive template
+     * @param template XML template with filled fields
+     */
+    void sendTemplateToRecipientUrlWithBA(HL7Recipient recipient, String template) throws IOException;
```
Comment (sebbrudzinski) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/service/IHETemplateService.java]: Do not use abbreviations in method names, if they are not well known (BA)

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImpl.java [hunk 3]
```diff
@@ -30,4 +32,21 @@ public void sendTemplateToRecipientUrl(String url, String template) throws IOExc
...
+        UsernamePasswordCredentials credentials = new UsernamePasswordCredentials(recipient.getRecipientUsername(),
+                recipient.getRecipientPassword());
+
+        post.setRequestEntity(new StringRequestEntity(template, "application/xml", "utf-8"));
+        client.getParams().setAuthenticationPreemptive(true);
+        client.getState().setCredentials(AuthScope.ANY, credentials);
+
+        LOGGER.info("Sending template with Basic Authentication to URL: {}", recipient.getRecipientUrl());
+        int responseCode = client.executeMethod(post);
+        LOGGER.info("Response code: {}", responseCode);
```
Comment (sebbrudzinski) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImpl.java:54]: It seems like a lot of this is repeated from the method above. Can we have a private method(s) for the common code?

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelper.java [hunk 4]
```diff
@@ -74,6 +74,11 @@ public void handleAction(Map<String, Object> parameters) throws IOException, Cla
         Template template = streamingTemplateEngine.createTemplate(new String(ArrayUtils.toPrimitive(templateData)));
         Writable writable = template.make(parameters);
         LOGGER.info("Template with name {}:\n{}", cdaTemplate.getTemplateName(), writable.toString());
-        iheTemplateService.sendTemplateToRecipientUrl(hl7Recipient.getRecipientUrl(), writable.toString());
+
+        if (hl7Recipient.getRecipientUsername() != null && hl7Recipient.getRecipientPassword() != null) {
```
Comment (sebbrudzinski) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/handler/helper/IHEActionHelper.java:78]: Can we test this logic in unit test?

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImpl.java [hunk 5]
```diff
@@ -30,4 +32,21 @@ public void sendTemplateToRecipientUrl(String url, String template) throws IOExc
         int responseCode = client.executeMethod(post);
         LOGGER.info("Response code: {}", responseCode);
     }
+
+    @Override
+    public void sendTemplateToRecipientUrlWithBA(HL7Recipient recipient, String template) throws IOException {
```
Comment (sebbrudzinski) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImpl.java]: Can we add unit test for this?

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/HL7Recipient.java [hunk 6]
```diff
@@ -17,6 +21,12 @@ public HL7Recipient(String recipientName, String recipientUrl) {
         this.recipientUrl = recipientUrl;
     }
 
+    public HL7Recipient(String recipientName, String recipientUrl, String recipientUsername, String recipientPassword) {
+        this(recipientName, recipientUrl);
```
Comment (sebbrudzinski) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/HL7Recipient.java]: The two args constructors should call this one. Not the other way.

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImpl.java [hunk 7]
```diff
@@ -19,14 +21,41 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger(IHETemplateServiceImpl.class);
 
+    private PostMethod post;
+
+    private HttpClient client;
```
Comment (sebbrudzinski) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImpl.java]: Why do we keep those two as class fields, if they are recreated with every method call we make? I don't see a reason they need to live after the method is executed.

**Comment by sebbrudzinski** [comment]
@mkruszynski I left some more comments

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImpl.java (hunk 7 above)
Comment (mkruszynski) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImpl.java]: I thought that these objects are used in several methods so they should be a class fields. If this is inappropriate, I'll change it.

Comment (sebbrudzinski) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/service/impl/IHETemplateServiceImpl.java]: Maybe I'm missing something, but I see two methods here that immadiately call "setConnectionParameters", which recreates the PostMethod and HttpClient from scratch.

**Comment by sebbrudzinski** [comment]
test this please

File: ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/HL7Recipient.java [hunk 8]
```diff
@@ -9,12 +9,22 @@
...
 
-    public HL7Recipient(String recipientName, String recipientUrl) {
+    public HL7Recipient(String recipientName, String recipientUrl, String recipientUsername, String recipientPassword) {
         this.recipientName = recipientName;
         this.recipientUrl = recipientUrl;
+        this.recipientUsername = recipientUsername;
+        this.recipientPassword = recipientPassword;
+    }
+
+    public HL7Recipient(String recipientName, String recipientUrl) {
```
Comment (sebbrudzinski) [inline ihe-interop/src/main/java/org/motechproject/ihe/interop/domain/HL7Recipient.java]: The constructors are usually ordered by the growing number of arguments they take

**Comment by sebbrudzinski** [comment]
test this please

**Pull request merged**